File: sdks/dotnet/src/DoowTrack/Management.cs
using System.Net.Http.Json;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace DoowTrack;

public class ManagementOptions
{
    public string Endpoint { get; init; } = "https://api.doow.co";
    public int TimeoutMs { get; init; } = 30000;
    public int RetryCount { get; init; } = 3;
}

public class Management : IDisposable
{
    private readonly string _apiKey;
    private readonly ManagementOptions _options;
    private readonly HttpClient _httpClient;
    private readonly JsonSerializerOptions _jsonOptions;

    public Management(string apiKey, ManagementOptions? options = null)
    {
        _apiKey = Environment.GetEnvironmentVariable("DOOW_TRACK_API_KEY") ?? apiKey;
        _options = options ?? new ManagementOptions();

        if (string.IsNullOrEmpty(_apiKey) || !_apiKey.StartsWith("dk_"))
        {
            throw new DoowError("Invalid API key format. Must start with 'dk_'.");
        }

        _httpClient = new HttpClient
        {
            Timeout = TimeSpan.FromMilliseconds(_options.TimeoutMs)
        };
        _httpClient.DefaultRequestHeaders.Add("Authorization", $"Bearer {_apiKey}");

        _jsonOptions = new JsonSerializerOptions
        {
            PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower,
            DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull
        };
        _jsonOptions.Converters.Add(new JsonStringEnumConverter(JsonNamingPolicy.SnakeCaseUpper));
    }

    public AppsResource Apps() => new(this);
    public ContractsResource Contracts() => new(this);
    public LicensesResource Licenses() => new(this);
    public MetricsResource Metrics() => new(this);
    public ExpensesResource Expenses() => new(this);

    internal async Task<T> RequestAsync<T>(HttpMethod method, string path, object? body = null)
    {
        var url = $"{_options.Endpoint.TrimEnd('/')}/sdk{path}";

        for (int attempt = 0; attempt <= _options.RetryCount; attempt++)
        {
            try
            {
                var request = new HttpRequestMessage(method, url);

                if (body != null)
                {
                    var json = JsonSerializer.Serialize(body, _jsonOptions);
                    request.Content = new StringContent(json, Encoding.UTF8, "application/json");
                }

                var response = await _httpClient.SendAsync(request);

                if (response.IsSuccessStatusCode)
                {
                    var content = await response.Content.ReadAsStringAsync();
                    return JsonSerializer.Deserialize<T>(content, _jsonOptions)!;
                }

                if ((int)response.StatusCode >= 500 && attempt < _options.RetryCount)
                {
                    await Task.Delay((int)Math.Pow(2, attempt) * 1000);
                    continue;
                }

                var errorBody = await response.Content.ReadAsStringAsync();
                throw new DoowError($"API error: {errorBody}", (int)response.StatusCode);
            }
            catch (HttpRequestException) when (attempt < _options.RetryCount)
            {
                await Task.Delay((int)Math.Pow(2, attempt) * 1000);
            }
        }

        throw new DoowError("Max retries exceeded");
    }

    internal async Task DeleteAsync(string path)
    {
        var url = $"{_options.Endpoint.TrimEnd('/')}/sdk{path}";

        for (int attempt = 0; attempt <= _options.RetryCount; attempt++)
        {
            try
            {
                var response = await _httpClient.DeleteAsync(url);

                if (response.IsSuccessStatusCode || response.StatusCode == System.Net.HttpStatusCode.NoContent)
                {
                    return;
                }

                if ((int)response.StatusCode >= 500 && attempt < _options.RetryCount)
                {
                    await Task.Delay((int)Math.Pow(2, attempt) * 1000);
                    continue;
                }

                var errorBody = await response.Content.ReadAsStringAsync();
                throw new DoowError($"API error: {errorBody}", (int)response.StatusCode);
            }
            catch (HttpRequestException) when (attempt < _options.RetryCount)
            {
                await Task.Delay((int)Math.Pow(2, attempt) * 1000);
            }
        }
    }

    public void Dispose()
    {
        _httpClient.Dispose();
        GC.SuppressFinalize(this);
    }
}

public class AppsResource
{
    private readonly Management _mgmt;
    internal AppsResource(Management mgmt) => _mgmt = mgmt;

    public async Task<PaginatedResponse<App>> ListAsync(string? cursor = null, int limit = 50)
    {
        var path = $"/apps?limit={limit}";
        if (!string.IsNullOrEmpty(cursor)) path += $"&cursor={cursor}";
        return await _mgmt.RequestAsync<PaginatedResponse<App>>(HttpMethod.Get, path);
    }

    public async Task<App> GetAsync(string id) =>
        await _mgmt.RequestAsync<App>(HttpMethod.Get, $"/apps/{id}");

    public async Task<App> CreateAsync(CreateAppInput input) =>
        await _mgmt.RequestAsync<App>(HttpMethod.Post, "/apps", input);

    public async Task<App> UpdateAsync(string id, UpdateAppInput input) =>
        await _mgmt.RequestAsync<App>(HttpMethod.Patch, $"/apps/{id}", input);

    public async Task DeleteAsync(string id) =>
        await _mgmt.DeleteAsync($"/apps/{id}");
}

public class ContractsResource
{
    private readonly Management _mgmt;
    internal ContractsResource(Management mgmt) => _mgmt = mgmt;

    public async Task<PaginatedResponse<Contract>> ListAsync(string appId, string? cursor = null, int limit = 50)
    {
        var path = $"/apps/{appId}/contracts?limit={limit}";
        if (!string.IsNullOrEmpty(cursor)) path += $"&cursor={cursor}";
        return await _mgmt.RequestAsync<PaginatedResponse<Contract>>(HttpMethod.Get, path);
    }

    public async Task<Contract> GetAsync(string id) =>
        await _mgmt.RequestAsync<Contract>(HttpMethod.Get, $"/contracts/{id}");

    public async Task<Contract> CreateAsync(string appId, CreateContractInput input) =>
        await _mgmt.RequestAsync<Contract>(HttpMethod.Post, $"/apps/{appId}/contracts", input);

    public async Task DeleteAsync(string id) =>
        await _mgmt.DeleteAsync($"/contracts/{id}");
}

public class LicensesResource
{
    private readonly Management _mgmt;
    internal LicensesResource(Management mgmt) => _mgmt = mgmt;

    public async Task<PaginatedResponse<License>> ListAsync(string? cursor = null, int limit = 50)
    {
        var path = $"/licenses?limit={limit}";
        if (!string.IsNullOrEmpty(cursor)) path += $"&cursor={cursor}";
        return await _mgmt.RequestAsync<PaginatedResponse<License>>(HttpMethod.Get, path);
    }

    public async Task<License> GetAsync(string id) =>
        await _mgmt.RequestAsync<License>(HttpMethod.Get, $"/licenses/{id}");
}

public class MetricsResource
{
    private readonly Management _mgmt;
    internal MetricsResource(Management mgmt) => _mgmt = mgmt;

    public async Task<PaginatedResponse<Metric>> ListAsync(string licenseId, string? cursor = null, int limit = 50)
    {
        var path = $"/licenses/{licenseId}/metrics?limit={limit}";
        if (!string.IsNullOrEmpty(cursor)) path += $"&cursor={cursor}";
        return await _mgmt.RequestAsync<PaginatedResponse<Metric>>(HttpMethod.Get, path);
    }

    public async Task<Metric> GetAsync(string id) =>
        await _mgmt.RequestAsync<Metric>(HttpMethod.Get, $"/metrics/{id}");

    public async Task<Metric> CreateAsync(string licenseId, CreateMetricInput input) =>
        await _mgmt.RequestAsync<Metric>(HttpMethod.Post, $"/licenses/{licenseId}/metrics", input);

    public async Task DeleteAsync(string id) =>
        await _mgmt.DeleteAsync($"/metrics/{id}");
}

public class ExpensesResource
{
    private readonly Management _mgmt;
    internal ExpensesResource(Management mgmt) => _mgmt = mgmt;

    public async Task<PaginatedResponse<Expense>> ListAsync(string appId, string? cursor = null, int limit = 50)
    {
        var path = $"/apps/{appId}/expenses?limit={limit}";
        if (!string.IsNullOrEmpty(cursor)) path += $"&cursor={cursor}";
        return await _mgmt.RequestAsync<PaginatedResponse<Expense>>(HttpMethod.Get, path);
    }
}

File: sdks/dart/lib/src/management.dart
import 'dart:convert';
import 'dart:io';
import 'package:http/http.dart' as http;
import 'types.dart';
import 'errors.dart';

class ManagementOptions {
  final String endpoint;
  final Duration timeout;
  final int retryCount;

  const ManagementOptions({
    this.endpoint = 'https://api.doow.co',
    this.timeout = const Duration(seconds: 30),
    this.retryCount = 3,
  });
}

class Management {
  final String _apiKey;
  final ManagementOptions _options;
  late final AppsClient apps;
  late final ContractsClient contracts;
  late final LicensesClient licenses;
  late final MetricsClient metrics;
  late final ExpensesClient expenses;

  Management(String apiKey, [ManagementOptions? options])
      : _apiKey = _resolveApiKey(apiKey),
        _options = options ?? const ManagementOptions() {
    if (!_apiKey.startsWith('dk_')) {
      throw ValidationError('API key must start with dk_');
    }
    apps = AppsClient(this);
    contracts = ContractsClient(this);
    licenses = LicensesClient(this);
    metrics = MetricsClient(this);
    expenses = ExpensesClient(this);
  }

  static String _resolveApiKey(String apiKey) {
    return Platform.environment['DOOW_TRACK_API_KEY'] ?? apiKey;
  }

  Future<Map<String, dynamic>> _request(
    String method,
    String path, {
    Map<String, dynamic>? body,
  }) async {
    final uri = Uri.parse('${_options.endpoint}$path');
    final headers = {
      'Authorization': 'Bearer $_apiKey',
      'Content-Type': 'application/json',
    };

    for (var attempt = 0; attempt <= _options.retryCount; attempt++) {
      try {
        http.Response response;
        switch (method) {
          case 'GET':
            response = await http.get(uri, headers: headers).timeout(_options.timeout);
            break;
          case 'POST':
            response = await http
                .post(uri, headers: headers, body: body != null ? jsonEncode(body) : null)
                .timeout(_options.timeout);
            break;
          case 'DELETE':
            response = await http.delete(uri, headers: headers).timeout(_options.timeout);
            break;
          default:
            throw DoowError('Unknown method: $method');
        }

        if (response.statusCode >= 200 && response.statusCode < 300) {
          if (response.body.isEmpty) return {};
          return jsonDecode(response.body) as Map<String, dynamic>;
        }

        if (response.statusCode == 401) {
          throw AuthError('Invalid API key');
        }

        if (response.statusCode == 429) {
          final retryAfter = response.headers['retry-after'];
          final delay = retryAfter != null
              ? Duration(seconds: int.tryParse(retryAfter) ?? 1)
              : Duration(milliseconds: 100 * (1 << attempt));
          await Future.delayed(delay);
          continue;
        }

        if (response.statusCode >= 500) {
          await Future.delayed(Duration(milliseconds: 100 * (1 << attempt)));
          continue;
        }

        throw DoowError('HTTP ${response.statusCode}: ${response.body}',
            statusCode: response.statusCode);
      } catch (e) {
        if (e is DoowError) rethrow;
        if (attempt == _options.retryCount) {
          throw DoowError('Request failed: $e');
        }
        await Future.delayed(Duration(milliseconds: 100 * (1 << attempt)));
      }
    }
    throw DoowError('Max retries exceeded');
  }
}

class AppsClient {
  final Management _mgmt;
  AppsClient(this._mgmt);

  Future<List<App>> list() async {
    final data = await _mgmt._request('GET', '/sdk/apps');
    final list = data['apps'] ?? data['data'] ?? [];
    return (list as List).map((e) => App.fromJson(e)).toList();
  }

  Future<App> get(String id) async {
    final data = await _mgmt._request('GET', '/sdk/apps/$id');
    return App.fromJson(data);
  }

  Future<App> create(CreateAppInput input) async {
    final data = await _mgmt._request('POST', '/sdk/apps', body: input.toJson());
    return App.fromJson(data);
  }

  Future<void> delete(String id) async {
    await _mgmt._request('DELETE', '/sdk/apps/$id');
  }
}

class ContractsClient {
  final Management _mgmt;
  ContractsClient(this._mgmt);

  Future<List<Contract>> list(String appId) async {
    final data = await _mgmt._request('GET', '/sdk/apps/$appId/contracts');
    final list = data['contracts'] ?? data['data'] ?? [];
    return (list as List).map((e) => Contract.fromJson(e)).toList();
  }

  Future<Contract> get(String id) async {
    final data = await _mgmt._request('GET', '/sdk/contracts/$id');
    return Contract.fromJson(data);
  }

  Future<Contract> create(String appId, CreateContractInput input) async {
    final data = await _mgmt._request('POST', '/sdk/apps/$appId/contracts',
        body: input.toJson());
    return Contract.fromJson(data);
  }

  Future<void> delete(String id) async {
    await _mgmt._request('DELETE', '/sdk/contracts/$id');
  }
}

class LicensesClient {
  final Management _mgmt;
  LicensesClient(this._mgmt);

  Future<List<License>> list(String contractId) async {
    final data = await _mgmt._request('GET', '/sdk/contracts/$contractId/licenses');
    final list = data['licenses'] ?? data['data'] ?? [];
    return (list as List).map((e) => License.fromJson(e)).toList();
  }

  Future<License> get(String id) async {
    final data = await _mgmt._request('GET', '/sdk/licenses/$id');
    return License.fromJson(data);
  }
}

class MetricsClient {
  final Management _mgmt;
  MetricsClient(this._mgmt);

  Future<List<Metric>> list(String licenseId) async {
    final data = await _mgmt._request('GET', '/sdk/licenses/$licenseId/metrics');
    final list = data['metrics'] ?? data['data'] ?? [];
    return (list as List).map((e) => Metric.fromJson(e)).toList();
  }

  Future<Metric> get(String id) async {
    final data = await _mgmt._request('GET', '/sdk/metrics/$id');
    return Metric.fromJson(data);
  }

  Future<Metric> create(String licenseId, CreateMetricInput input) async {
    final data = await _mgmt._request('POST', '/sdk/licenses/$licenseId/metrics',
        body: input.toJson());
    return Metric.fromJson(data);
  }

  Future<void> delete(String id) async {
    await _mgmt._request('DELETE', '/sdk/metrics/$id');
  }
}

class ExpensesClient {
  final Management _mgmt;
  ExpensesClient(this._mgmt);

  Future<List<Expense>> list() async {
    final data = await _mgmt._request('GET', '/sdk/expenses');
    final list = data['expenses'] ?? data['data'] ?? [];
    return (list as List).map((e) => Expense.fromJson(e)).toList();
  }

  Future<Expense> create(CreateExpenseInput input) async {
    final data = await _mgmt._request('POST', '/sdk/expenses', body: input.toJson());
    return Expense.fromJson(data);
  }
}
